[CURL] Add support for data url link downloads

yury-s opened the pull request:
#### 8050ec39e480cea3e8046942831a6edc31bf17d1
<pre>
[CURL] Add support for data url link downloads
<a href="https://bugs.webkit.org/show_bug.cgi?id=255696">https://bugs.webkit.org/show_bug.cgi?id=255696</a>

Reviewed by Fujii Hironori.

Curl does not support data: urls, handle them using
NetworkDataTaskDataURL.

* LayoutTests/platform/wincairo/TestExpectations:
* Source/WebCore/platform/network/DataURLDecoder.h:
* Source/WebKit/NetworkProcess/NetworkDataTask.cpp:
* Source/WebKit/PlatformPlayStation.cmake

Canonical link: <a href="https://commits.webkit.org/263434@main">https://commits.webkit.org/263434@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2519f6f259356569861004d2d7cf46fabce6964e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4567 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4687 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4838 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/6058 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4727 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/4558 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4829 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4643 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/4965 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4635 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4736 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/4089 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/6064 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/2228 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/4081 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/9077 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/4084 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/4153 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/5692 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4544 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3705 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4077 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1133 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8118 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/4438 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kshukuwa commented on Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.cpp:
```diff
@@ -84,12 +84,18 @@ NetworkDataTaskCurl::NetworkDataTaskCurl(NetworkSession& session, NetworkDataTas
+    if (request.url().protocolIsData()) {
+        DataURLDecoder::decode(request.url(), { }, [this, protectedThis = Ref { *this }](auto decodeResult) mutable {
+            didReadDataURL(WTFMove(decodeResult));
+        });
```
This process should execute at NetworkDataTaskCurl::resume(). Also, it should check m_state before execute didReadDataURL().

kshukuwa commented on Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.cpp:
```diff
@@ -283,6 +289,36 @@ bool NetworkDataTaskCurl::shouldRedirectAsGET(const ResourceRequest& request, bo
     return false;
 }
 
+void NetworkDataTaskCurl::didReadDataURL(std::optional<DataURLDecoder::Result>&& result)
```
It seems that there are cases where result is not returned.

https://github.com/WebKit/WebKit/blob/d488f398f95d05b533bb5a45084b49d91c36ee84/Source/WebKit/NetworkProcess/soup/NetworkDataTaskSoup.cpp#L1802-L1805

kshukuwa replied:
Ah, you are checking with downloadDataURL. However, since it is useless to execute didReceiveResponse, is it possible to execute failure processing inside this function ? NetworkDataTaskSoup::didFailDownload may be helpful.

yury-s replied:
Yes, I can certainly move it.

kshukuwa commented on Source/WebKit/PlatformWin.cmake:
```diff
@@ -19,6 +19,8 @@ list(APPEND WebKit_SOURCES
     GPUProcess/win/GPUProcessMainWin.cpp
     GPUProcess/win/GPUProcessWin.cpp
 
+    NetworkProcess/NetworkDataTaskDataURL.cpp
```
Could you add the same line to Source/WebKit/PlatformPlayStation.cmake as well? They also use CURL.

fujii commented on LayoutTests/platform/wincairo/TestExpectations:
```diff
@@ -936,7 +936,7 @@ http/tests/security/XFrameOptions/x-frame-options-deny.html [ Skip ] # Failure
 http/tests/security/XFrameOptions/x-frame-options-multiple-headers-sameorigin-deny.html [ Skip ] # Failure
 http/tests/security/XFrameOptions/x-frame-options-parent-same-origin-deny.html [ Skip ] # Failure
-http/tests/security/anchor-download-allow-data.html [ Skip ] # Crash
+http/tests/security/anchor-download-allow-data.html [ Pass ]
```
I think we can just remove this line.

fujii reviewed (approved):
LGTM. Thank you very much.

yury-s committed:
[CURL] Add support for data url link downloads
https://bugs.webkit.org/show_bug.cgi?id=255696

Reviewed by Fujii Hironori.

Curl does not support data: urls, handle them using
NetworkDataTaskDataURL.

* LayoutTests/platform/wincairo/TestExpectations:
* Source/WebCore/platform/network/DataURLDecoder.h:
* Source/WebKit/NetworkProcess/NetworkDataTask.cpp:
* Source/WebKit/PlatformPlayStation.cmake

Canonical link: https://commits.webkit.org/263434@main
